### PR TITLE
sys/tasklet: rename irq_handler to tasklet

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR is a replacement of #12420. It renames the `irq_handler` module (#10555) to `tasklet`, as described in https://github.com/RIOT-OS/RIOT/pull/12420#issuecomment-541015981

The reason is because the concept of "tasklets" is common in OSs and scheduling systems, so this way it should be easier to expose this feature in the documentation.

### Testing procedure
Use `tests/sys_tasklet` as following.
```
make -C tests/sys_tasklet BOARD=...
```

### Issues/PRs references
#10555 
Closes #12420

### CODE REVIEW

**Comment by jia200x** [comment]
hi @kaspar030

Note tasklets are not a replacement of event queues, they are used for different things. The irq_handler implementation (renamed here to tasklets) is close to the Tasklets implementation of Linux and the event API closer to the Workqueues of Linux (check the differences here https://0xax.gitbooks.io/linux-insides/content/Interrupts/linux-interrupts-9.html).

Using event loops on irq_handler/tasklets is a design decision, same as using softIRQ for implementing tasklets in Linux.

> it adds:
    thread creation function
    a check on every event post whether the thread is running
    a check whether a handler has already been registered ("pending flag")

I'm aware of this, and this was not the case of the original tasklets PR ( #12420 ).
I also understand @gschorcht argument (https://github.com/RIOT-OS/RIOT/pull/12420#issuecomment-541014962) that it doesn't waste memory the `irq_handler_add` function is not called, but on the other hand the module wouldn't be included if not used.

I'm with @kaspar030 on this argument, but this can be modified to the behavior of #12420 

> It removes the ability to have multiple queues (there's only one highest prio thread running a single queue), and it's API description makes it seem like a good way to defer ISR work to thread context.

It doesn't remove the ability to have multiple event queues. The `event` module is used for that.
We might need an extra "tasklet" thread (TASKLET_HI) if we need preemptiveness for tasklets.

> 3.) the pending flag is IMO implicit in event, as event->next is always set when in queue (pending), and NULL when not. If not, this could have been added to event.

True. Already solved in #12420, and can be adapted here as well.

> Now you're proposing to rename this wrapper around event to "tasklet", which seems to some to be a well-known name for "put function plus argument on a queue, have a loop collect and execute it later".
At that point, when would I use tasklets, when would I use events?

Tasklets should be used for handling low level stuff (PHY/MAC, drivers ISR offloading, internal OS signals, etc). In order to preserve real time constrains, they should aim to be deterministic.

Events should be used for all the rest (interfacing to user API, network stack components, libraries that require timeouts, and in most cases as a replacement of IPC messages).

> IMO, they're the same. "irq_handler" (or "tasklets") have some ISR related how-to-use documentation, and there's a shared thread for executing the handlers. But that one thread might not be enough has already been discussed. So has adding shared event-threads for the event module.

The tasklets wrap the `event` module. But, the `event` loop relays on a thread given by the user of the module. A tasklet can be scheduled from everywhere without caring about who process the events

> I think we should consolidate to one "event / event loop" API.

If that means that the `event` api adds the possibility to process events from the OS (so adding one or more threads to process this events) then I would be fine. Otherwise, network devices, drivers and different modules might need to add their own thread to process their events (as described by @gschorcht in the motivation of the `irq_handler`)

**Comment by kaspar030** [comment]
> But, the `event` loop relays on a thread given by the user of the module. A tasklet can be scheduled from anywhere without caring about who process the events

Yes, because there's no "shared event thread" within the current event module. If there were, instead of
`tasklet_schedule(tasklet)` you'd use `event_post(SHARED_QUEUE, my_event)`. What I'm saying is, if "event" doesn't have the shared thread(s), adding irq_handler + shared thread, and then argue irq_handler is needed because it has shared threads and event has not, doesn't make sense if they're otherwise identical.

Let's add shared threads to event (as @haukepetersen already has, unfortunately hidden in a branch).

> Tasklets should be used for handling low level stuff (PHY/MAC, drivers ISR offloading, internal OS signals, etc). In order to preserve real time constrains, they should aim to be deterministic.
> 
> Events should be used for all the rest (interfacing to user API, network stack components, libraries that require timeouts, and in most cases as a replacement of IPC messages).

How do they differ (other than being put in different queues)?

**Comment by gschorcht** [comment]
> 2.) Has to be removed. ... Worse, if the thread has not been started, it will, from ISR context, at less than defined time. (It might happen on the first radio packet received...). Thus that has to be moved into an initialization function.

The single interrupt handler thread was initially motivated by the fact that there are a lot of drivers for sensors that are connected via I2C or SPI and the mutex based synchronization mechanism doesn't work in ISRs.

The question at that time was, who should create the handler thread. The different drivers have no knowledge about the handler thread. The application should't need any knowledge about the internals of the drivers and whether they need the handler thread or not. That was the reason why the thread was created when it is used the first time.

A possible solution could be the autoinitialization if `irq_handler` module is enabled.

**Comment by jia200x** [comment]
> Yes, because there's no "shared event thread" within the current event module. If there were, instead of
tasklet_schedule(tasklet) you'd use event_post(SHARED_QUEUE, my_event). What I'm saying is, if "event" doesn't have the shared thread(s), adding irq_handler + shared thread, and then argue irq_handler is needed because it has shared threads and event has not, doesn't make sense if they're otherwise identical.

What you describe is exactly the same as a tasklet, but it's explicitly called "tasklet" instead of "shared_queue". The problem here is only about semantics, because the "shared queue" would also need to initialize the thread, put the queue somewhere, etc.

So, this is only a naming convention to me.

> How do they differ (other than being put in different queues)?

Semantics. Tasklets are intended to be used for "Bottom Half Processing" or deferring the execution until the kernel finds a safe time to run the task.
Events on the other hand, are a generic mechanism of triggering the execution of a task in the context of the queue.

Tasklets is just an implementation of this "shared event thread" that shares the definition of tasklets of most libraries and OSs.

**Comment by kaspar030** [comment]
> Semantics. Tasklets are intended to be used for "Bottom Half Processing" or deferring the execution until the kernel finds a safe time to run the task.
> Events on the other hand, are a generic mechanism of triggering the execution of a task in the context of the queue.

How are they semantically different? Not what the name implies to you what they should be used for, but how a function executed as a tasklet is semantically different from a function executed as an event handler?

IIUC, apart from naming, they are equivalent.

> Tasklets is just an implementation of this "shared event thread" that shares the definition of tasklets of most libraries and OSs.

I understand that Linux and Zephyr call their function deferring mechanism "tasklet". Where else?
IMO, especially application developers will look for libevent/libev/libuv style event queues.

**Comment by jia200x** [comment]
> How are they semantically different? Not what the name implies to you what they should be used for, but how a function executed as a tasklet is semantically different from a function executed as an event handler?
IIUC, apart from naming, they are equivalent.

I meant language semantics. Since tasklets/irq_handler are implemented on top of events, they are the same in terms of functionality.
As said, the tasklet is just the name of this "shared thread" queue.

> I understand that Linux and Zephyr call their function deferring mechanism "tasklet". Where else?

OpenThread, Spring Batch, some Google JS libraries, etc.

> IMO, especially application developers will look for libevent/libev/libuv style event queues.

Exactly, and they should use events instead of tasklets. As said before, tasklets are intended to be used by the kernel components. I don't see why an end user should offload an ISR or do something really low level if there's a driver that handles that. If someone needs low level operations (even on Linux), then would be closer working closer to the kernel.

**Comment by kaspar030** [comment]
> Since tasklets/irq_handler are implemented on top of events, they are the same in terms of functionality. As said, the tasklet is just the name of this "shared thread" queue.

> Exactly, and they should use events instead of tasklets. As said before, tasklets are intended to be used by the kernel components. I don't see why an end user should offload an ISR or do something really low level if there's a driver that handles that. If someone needs low level operations (even on Linux), then would be working closer to the kernel.

Is your point that because some deferred work is "low level", "Bottom Half" or "ISR offload", it should be done by an API named "tasklet" because that is a commonly used name for that, and if it is "application stuff", it should be done with a differently named API because "tasklets" is not supposed to be used for that?

**Comment by haukepetersen** [comment]
Hej, I'd like to jump into this discussion, too - as I seem to have missed the discussion when `irq_handler` was merged...

I have to say that I get a bad stomachache looking at the concept. My major concern is that it **destroys all real-time properties** for device drivers that use this thread! In the current form it is highly dependent on the selected modules on how they will perform. Example: a none-time-critical sensor on a slow I2C bus can block a highly time critical accelerometer/radio/whatever on a completely independent SPI bus -> bad, right!?

Taking a step back: the general concept of creating a shared thread context that can be used by multiple modules is a very nice thing to do, especially considering the benefits in saved system resources (stack space, ...). So I am in for that.  BUT: just having a single thread for **various** interrupt operations is IMHO not an option (see example above)!

So my proposal would be something like the following:
- have a more generic module that allows (via compile time configuration) to spawn any number of event handler threads with configurable priorities. 
- allow to configure (at compile time) which handler thread to actually use for each driver/module that wants to use one

So in a default configuration, I could imagine one high-prio handler for drivers, and maybe a low-prio handler for logging and similar. So in that case, the approach would not differ much from what is proposed here. BUT: with that approach it is possible to implement more fine grained setups if needed!

**Comment by bergzand** [comment]
> I have to say that I get a bad stomachache looking at the concept. My major concern is that it **destroys all real-time properties** for device drivers that use this thread! In the current form it is highly dependent on the selected modules on how they will perform. Example: a none-time-critical sensor on a slow I2C bus can block a highly time critical accelerometer/radio/whatever on a completely independent SPI bus -> bad, right!?

I think this can be extended to any high priority thread getting blocked due to a number of slow and unpredictable I2C bus transaction taking priority. This is the strong point of the current model where interrupts set only a flag and the IRQ itself is handled at the priority of the handling thread (see below). With the current design having priority 0 as a must the timing behaviour for other threads becomes rather unpredictable¸ being dependent on bus usage among other things.

What I'm afraid of here is that a method is made available to schedule relative slow processes (I2C bus exchanges with a potential timeout or clock stretching) on a high priority thread. This to me looks like an easy and not so transparant way to hamper other relative high priority threads such as netif threads. Which could show itself as a non-deterministic RTT spike in a network packet.

> So my proposal would be something like the following:
> 
>     * have a more generic module that allows (via compile time configuration) to spawn any number of event handler threads with configurable priorities.
> 
>     * allow to configure (at compile time) which handler thread to actually use for each driver/module that wants to use one

To me it would already help if this tasklets idea can be modified to run as a low priority thread where tasks are run eventually.

Background:

**Comment by gschorcht** [comment]
I think we all agree that it would be nice to have something that helps to handle events from different drivers that require the access to exclusive resources in thread context, without having to create a separate thread for each driver.

The question is how to realize and how it should be configured.

**Comment by gschorcht** [comment]
> So when introducing shared event handlers, I think it is very important to leave that door open to developers, so they still have the power to fine tune their system regarding runtime priorities.

Just for clarification, the priority of the `irq_handler` is configurable at the compile time.

**Comment by bergzand** [comment]
> Just for clarification, the priority of the `irq_handler` is configurable at the compile time.

Technically yes, but if the [comment here](https://github.com/RIOT-OS/RIOT/pull/12459/files#diff-1e1b1798d3c8f1b04b8d495dd227ed1fR170-R174) is to be trusted there isn't that much to configure.

**Comment by haukepetersen** [comment]
> I think we all agree that it would be nice to have something that helps to handle events from different drivers that require the access to exclusive resources in thread context, without having to create a separate thread for each driver.

Absolutely.

> The question is how to realize and how it should be configured.

In the core I think it can be pretty straight forward by running a number of event handler threads during system initialization, each simply running `event_loop()`. The configuration for this module could look similar to all the `x_params.h` constructs we use for drivers:
```c 
typedef struct {
    char priority;
    const char *name; // make this possibly optional
} abceventrunner_params_t;

static const abceventrunner_params_t abceventrunner_params[] =
{
    { 1, "irq_handler_high" },
    { 4, "irq_handler_med" },
    { (THREAD_PRIORITY_IDLE - 1), "run_me_if_you_can" },
}
```
Initialization is trivial, simply allocate `ARRAY_SIZE(abceventrunner_params)` stacks, and loop `thread_create()` over these... The thread function is identical for all of them, a one-liner calling `event_loop()` -> voila, a configurable event handler setup with user definable priorities.

I started to play with something like this a time ago: https://github.com/haukepetersen/RIOT/tree/add_eventhandlerthread, though never came to integrate multiple queues...

Now for drivers we just need to add some syntactic sugar so we can directly tell them which queue to use.

**Comment by haukepetersen** [comment]
> Just for clarification, the priority of the irq_handler is configurable at the compile time.

Yes, but only collectively without any option for differentiation (slow dev A vs picky dev B...)

**Comment by gschorcht** [comment]
> > Just for clarification, the priority of the `irq_handler` is configurable at the compile time.
> 
> Technically yes, but if the [comment here](https://github.com/RIOT-OS/RIOT/pull/12459/files#diff-1e1b1798d3c8f1b04b8d495dd227ed1fR170-R174) is to be trusted there isn't that much to configure.

Yes indeed, you're right :smile:

**Comment by kaspar030** [comment]
> Now for drivers we just need to add some syntactic sugar so we can directly tell them which queue to use.

Every event queue can be defined somehow:

`#define EVENT_PRIO_HIGHEST (&queue-pointer)`

Then it can be called like this, from the drivers:

````
#ifndef DRIVER_FOO_EVENT_QUEUE
#define DRIVER_FOO_EVENT_QUEUE EVENT_PRIO_HIGHEST
#endif

event_post(DRIVER_FOO_EVENT_QUEUE, event);
```

We can make existence of the queues available via pseudomodules.

**Comment by maribu** [comment]
But lets also keep in mind that RIOT cannot magically fulfill any real time capabilities for any use case out of the box, but only provide the tools to ease providing a real time capable app. E.g. if the app developer puts an fast I2C device and a notorious clock stretcher on the same bus - even `USEMODULE += bend_the_law_of_phyiscs` will likely not help... My point: We cannot reason about the hardware and software settings required to fulfill the real time requirements in a certain use case - but the app developer can.

So making it easy for the application programmer to assign the prios (e.g. like @kaspar030 suggested) and spawn any arbitrary number "work thread" with user-provided scheduling prios is in my opinion the next logical step in the development.

But trying to reason from our perspective which drivers have by default a higher or lower prio than others is something I advocate to avoid. I say just spawn a single thread with some default prio in absence of any config provided by the app developer. If the app developers wants real time capability, they simply has to provide their own config.

File: sys/include/tasklet.h [hunk 1]
```diff
@@ -162,92 +165,92 @@ extern "C" {
 #endif
 
 /**
- * @brief   Default priority of the interrupt handler thread
+ * @brief   Default priority of the tasklets thread
```
[the remark this answers is not in the document]
Comment (maribu) [reply sys/include/tasklet.h]: I advocate to not expose this now to a broader audience. While the API of offloading work to another thread seems to fulfill all needs, the configuration of the taskletthread is very likely to be changed in the near future to allow multiple threads, if requested.

**Comment by kaspar030** [comment]
I'd like to get some opinions on the differentiation between isr offloading kind of tasklets/events and any networking/application tasklets/events. Do we have to treat them differently? I mean, obviously applications might want to define their own event queues, and ISR events need to be handled by a different and/or dedicated thread, but do we need seperate APIs?

**Comment by haukepetersen** [comment]
@maribu that is exactly what I was trying to say: keep the default straight forward, BUT allow for elaborate setups.

@kaspar030 my answer is no. IMHO it would be even counter-productive to encapsulate, and thereby limit, the shared event handlers in any means. Providing these shared event handlers in the most generic fashion allows for perfect extensibility in the future and potentially some sweet optimization for any kind of module. 

E.g. modules that need an event handler could be made configurable to either spawn their own event handler thread, or to use a suitable shared one... Many many possibilities :-)

File: sys/include/tasklet.h [hunk 2]
```diff
@@ -162,92 +165,92 @@ extern "C" {
...
- * @brief   Interrupt event structure
+ * @brief   Tasklet structure
  *
- * Using modules have to define a structure of this type for each interrupt
- * source used by the modules. Structures of this type are used to put
- * them in a pending interrupt queue indicating that an interrupt of the
- * corresponding source has occurred and needs to be handled. Each interrupt
- * event can only be pending once.
+ * Used modules have to define a structure of this type for each tasklet
+ * used by the modules. Structures of this type are used to put
```
[the remark this answers is not in the document]
Comment (maribu) [reply sys/include/tasklet.h]: How about "Modules using tasklets have to define..."

File: sys/include/tasklet.h [hunk 3]
```diff
@@ -162,92 +165,92 @@ extern "C" {
...
 /**
- * @brief   Interrupt handling function prototype
+ * @brief   Tasklet handling function prototype
  *
  * Defines the prototype of the function that is registered together
- * with an interrupt event and to be called when the interrupt is handled.
+ * with a taskletand to be called when the tasklet is handled.
  */
-typedef void (*irq_isr_t)(void *ctx);
+typedef void (*tasklet_task_t)(void *ctx);
```
Comment (maribu) [inline sys/include/tasklet.h]: I'm not sure if "task" is the right term. To me, this is the signature of a function that is executing the actual task. But the function itself is not the task. Maybe a different term would be clearer?

File: sys/tasklet/tasklet.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
+/*
```
Comment (maribu) [inline sys/tasklet/tasklet.c]: This is not displayed as a rename by github. I didn't check if this is GitHubs fault, or if somehow in the commit this was not correctly detected as a rename.

Can you check if this is indeed not an issue with the commit, but only a bug in GitHub?

**Comment by haukepetersen** [comment]
Looking at the code in #12474, I must say its simplicity is quite intriguing. It is by no means perfect (I bet we can find something less 'ifdeffy' for the stacksize config). But it demonstrates quite nicely, that the `event` module does actually already provice all the features needed for the problem at hand.

I am strongly with @kaspar030 here, that I don't see a reason for augmenting an existing interface without actually adding any additional functionality - in contrary, adding overhead for the same functionality. Especially having the ISR offloading as primary use-case in mind, we are talking about critical code paths after all...

**Comment by haukepetersen** [comment]
Just played quickly with some ideas that were discussed here: #12480.

**Comment by maribu** [comment]
> I prefer to have concise modules that do only one thing (easier to test, easier to replace in the future) instead of extending a current one

In @kaspar030's PoC he is adding the module `event_thread` as a submodule of `event`. Does this count?

**Comment by haukepetersen** [comment]
> My original intention was because tasklet and event are not exactly the same

as was shown: from a technical perspective they are :-)

> Tasklets can e.g run in several cores

Lets not worry about multi-core systems (for now) -> these have so many implications on all the core modules (msg, mutex, ...), that event threads are the least concern...

> That's all I can say. If all here prefer to extend the event module I'm fine and we can close this PR.

+1 for going the `event module` path from my side. But how about to rework this PR in that direction instead of closing it?!

**Comment by jia200x** [comment]
> note 1: lets not worry about multi-core systems for now -> there are huge number of implications for all core modules with that, and the event handler threads are the least pressing...

Yes, I'm aware of that. I was only referring on why tasklets differ from event queues.

> Once we have multiple cores, starting multiple handler threads and changing event to use a mutex instead of thread_flags for notification would solve that. Once we have actual multicore support....

That's exactly the point why I was arguing for `tasklets/irq_handler/whatever` instead of extending the `event` module. You can change the implementation of the tasklets mechanism to semaphores, mutex, whatever at any time. 

What's the technical reason behind extending the API of `event` instead of having it's own module, considering the implementation of the `tasklet` module using `event` is one of many ways to implement it?

**Comment by jia200x** [comment]
> +1 for going the event module path from my side. But how about to rework this PR in that direction instead of closing it?!

I meant, if it's decided to go with the extension of the `event` module, it makes no sense to continue here if there are already 2 PoC PRs

**Comment by kaspar030** [comment]
> What's the technical reason behind extending the API of `event` instead of having it's own module, considering the implementation of the `tasklet` module using `event` is one of many ways to implement it?

The only extension that we need for now is the creation of dedicated handler threads. Otherwise, the event API is already a "tasklet" implementation that is already there. What would be the technical reason to create a new module, if an existing one provides all the functionality needed?

**Comment by kaspar030** [comment]
> That's exactly the point why I was arguing for `tasklets/irq_handler/whatever` instead of extending the `event` module. You can change the implementation of the tasklets mechanism to semaphores, mutex, whatever at any time.

events *are* tasklets. We can change the event implementation to semaphores, mutex, whatever at any time...

**Comment by jia200x** [comment]
> events are tasklets. We can change the event implementation to semaphores, mutex, whatever at any time...

This is the root of the problem. Tasklets and events are not equivalent.

1. Tasklets run ASAP as the OS finds a safe time and it doesn't matter which thread/core run them.
2. Tasklets may have priorities, but they still stick to point 1.

The low priority tasks introduced in #12480 and #12474 are not considered tasklets!

I'm aware tasklets can be implemented using `events`, but I insist they are not the same!

**Comment by jia200x** [comment]
> Tasklets and events are not equivalent.

On @haukepetersen comment above, yes. This PR and the current `irq_handler` on master don't reflect what I wrote above. But that's the aim of the `tasklet` concept.

**Comment by kaspar030** [comment]
> This is the root of the problem. Tasklets and events are not equivalent.

I think you're quite fixed on what tasklets are doing in Linux (which is quite a different environment).

>     1. Tasklets run ASAP as the OS finds a safe time and it doesn't matter which thread/core run them.

On Linux, Tasklets are *always* run on the CPU on which they were scheduled.
Same on RIOT when using events.
Depending on the priority of the worker thread, on RIOT, the event handler will be run ASAP as well, unless another event is currently being handled by that thread. Same as Tasklets on Linux.

>     2. Tasklets may have priorities, but they still stick to point 1.

So do events when using multiple handler threads.

**Comment by jia200x** [comment]
Ok, it was not the intention to make that a big discussion because of naming conventions or where to implement a module. I gave my reasons to implement it on top of the `event` loop and not next to because of maintainability. I don't want to stall the discussion neither.

If I'm the only one aiming for a `tasklet` module, then I respect the rough consensus approach, drop this PR and continue with #12474.

**Comment by maribu** [comment]
@gschorcht: Would you be fine if I open a PR targeting the release candidate that removes `irq_handler` for now? As the API is under heavy discussion, I think it would better to not get the feature to a broader audience yet. Otherwise we likely have to maintain two APIs.

**Comment by gschorcht** [comment]
> @gschorcht: Would you be fine if I open a PR targeting the release candidate that removes `irq_handler` for now? As the API is under heavy discussion, I think it would better to not get the feature to a broader audience yet. Otherwise we likely have to maintain two APIs.

Of course. It seems that we an agreement.

**Comment by bergzand** [comment]
I'm missing a bit of background information here. Is there an example of a sensor/module that requires processing ASAP (and thus where a low priority queue doesn't work) after the interrupt but where it can't be done in the IRQ. In other words, what's the use case of the high priority handler?

**Comment by gschorcht** [comment]
> I'm missing a bit of background information here. Is there an example of a sensor/module that requires processing ASAP (and thus where a low priority queue doesn't work) after the interrupt but where it can't be done in the IRQ. In other words, what's the use case of the high priority handler?

Each sensor that requires access to SPI or I2C after the sensor triggered an interrupt, for example when exhausting a defined threshold. Another example are I2C GPIO extenders that indicate the change of an input by an interrupt.

**Comment by bergzand** [comment]
> Each sensor that requires access to SPI or I2C after the sensor triggered an interrupt, for example when exhausting a defined threshold.

Maybe I'm missing the point here, but why does the average sensor require a **high priority** thread to handle a threshold exceeded notification?

>  Another example are I2C GPIO extenders that indicate the change of an input by an interrupt. The access to SPI and I2C is only possible in thread context.

This is IMHO the only example I've seen so far having a hard requirement on the concept discussed here. Just to have the option out in the open here: it is perfectly valid to decide not to support interrupt for these GPIO expander devices and only support GPIO interrupts on the MCU built-in GPIO.

**Comment by bergzand** [comment]
On a more constructive note, is it okay with people to as a first step create the concept discussed here as a low priority thread by default first. This gives us the option to validate the API and ideas discussed here in an unobtrusive way for the rest of the ecosystem. While it might not be suitable for the GPIO expander use case, I think it would still be suitable for sensor notification handling.

We can always extend to multiple handler threads and/or multiple priorities later if deemed necessary.

**Comment by gschorcht** [comment]
> > Each sensor that requires access to SPI or I2C after the sensor triggered an interrupt, for example when exhausting a defined threshold.
> 
> Maybe I'm missing the point here, but why does the average sensor require a **high priority** thread to handle a threshold exceeded notification?

It depends on the sensor. If it is just a temperature sensor, exceeding a threshold might not to be important. But what about a accelerometer in security critical applications? The system has to react in milliseconds.

**Comment by bergzand** [comment]
> > > Each sensor that requires access to SPI or I2C after the sensor triggered an interrupt, for example when exhausting a defined threshold.
> > 
> > 
> > Maybe I'm missing the point here, but why does the average sensor require a **high priority** thread to handle a threshold exceeded notification?
> 
> It depends on the sensor. If it is just a temperature sensor, exceeding a threshold might not to be important. But what about a accelerometer in security critical applications? The system has to react in milliseconds.

I wouldn't call an accelerometer in a security critical application the *average sensor*. Doesn't mean that your example is invalid, but I don't think it is the common case.

**Comment by maribu** [comment]
I think we have concluded that we let the user configure the priority levels, as arguing about priorities without the context of the intended application makes no sense, right?

But I agree with @bergzand that in absense of any user provided priority assignment low priority is reasonable default (maybe just higher than the main thread).

**Comment by gschorcht** [comment]
> low priority is reasonable default (maybe just higher than the main thread).

This is exactly what I already proposed in https://github.com/RIOT-OS/RIOT/pull/12474#discussion_r335972034 to avoid that drivers use the high priority thread to ensure to be handled before the main thread.

**Comment by gschorcht** [comment]
> While it might not be suitable for the GPIO expander use case, I think it would still be suitable for sensor notification handling.

@bergzand By the way, the need for a single handler thread came up when I wrote the driver for PFC857x in PR #10430. Therefore, the handler priority was originally high.

**Comment by jia200x** [comment]
> I think we have concluded that we let the user configure the priority levels, as arguing about priorities without the context of the intended application makes no sense, right?

But I agree with @bergzand that in absense of any user provided priority assignment low priority is reasonable default (maybe just higher than the main thread).

I think this went a little bit out of the context of my original motivation, although this approach could be useful for other things too.

My original motivation was to have an event queue with a "high enough" priority, so the OS can handle transceiver ISR and MAC layer logic. So, a low priority thread wouldn't work for my use case.

If low level priorities are useful for notifications, I think #12474 or #12480 are good candidates for both use cases

**Comment by gschorcht** [comment]
> My original motivation was to have an event queue with a "high enough" priority, so the OS can handle transceiver ISR and MAC layer logic. So, a low priority thread wouldn't work for my use case.

That's why PR #12474 defines a high priority and a low priority thread. That should cover both cases. IMHO, the question is how low should the low priority be, less than the main thread or less than the idle thread.

**Comment by kaspar030** [comment]
> @kaspar030: Are you willing to complete your proof of concept PR?

Sure!

**Comment by maribu** [comment]
> I think this went a little bit out of the context of my original motivation, although this approach could be useful for other things too.

How about this: We use something like `EVENT_HANDLER_PRIO_DEFAULT` and `EVENT_HANDLER_PRIO_DEFAULT_HIGH` for the default priors in the drivers, but enable the use to overwrite the priority of each event source individually. And we also also the user to overwrite whatever value those two default priorities have. That way we can provide a default that will work with the most common use cases out of the box.

I think it would also be a nice implementation detail, if the number of event handler threads could be chosen regardless of the actually used priorities. E.g. if only one event handler thread is enabled, that thread should just handle all events regardless of their priorities. The more threads are actually created, the more priority levels start to actually show different behavior.

**Comment by jia200x** [comment]
> How about this: We use something like EVENT_HANDLER_PRIO_DEFAULT and EVENT_HANDLER_PRIO_DEFAULT_HIGH for the default priors in the drivers, but enable the use to overwrite the priority of each event source individually. And we also also the user to overwrite whatever value those two default priorities have. That way we can provide a default that will work with the most common use cases out of the box.
I think it would also be a nice implementation detail, if the number of event handler threads could be chosen regardless of the actually used priorities. E.g. if only one event handler thread is enabled, that thread should just handle all events regardless of their priorities. The more threads are actually created, the more priority levels start to actually show different behavior.

That would be nice.
Then as proposed in https://github.com/RIOT-OS/RIOT/pull/12480#issuecomment-543634852 it would be possible to have default priorities for each module. So we can not only configure in which queue a module should have its function processed, but also what's the priority of that queue.

**Comment by jia200x** [comment]
I guess we can close this one since #12474 got merged.

**Pull request closed without merging**